File: analytics/metrics-platform/src/main/java/org/wikimedia/metrics_platform/EventSenderDefault.java
package org.wikimedia.metrics_platform;

import static java.util.logging.Level.INFO;

import java.io.IOException;
import java.net.URL;
import java.util.Collection;

import org.wikimedia.metrics_platform.event.EventProcessed;

import com.google.gson.Gson;

import lombok.extern.java.Log;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;
import okhttp3.ResponseBody;

@Log
public class EventSenderDefault implements EventSender {

    private final Gson gson;
    private final OkHttpClient httpClient;

    public EventSenderDefault(Gson gson, OkHttpClient httpClient) {
        this.gson = gson;
        this.httpClient = httpClient;
    }

    @Override
    public void sendEvents(URL baseUri, Collection<EventProcessed> events) throws IOException {
        Request request = new Request.Builder()
            .url(baseUri)
            .header("Accept", "application/json")
            .header("User-Agent", "Metrics Platform Client/Java " + MetricsClient.METRICS_PLATFORM_LIBRARY_VERSION)
            .post(RequestBody.create(gson.toJson(events), okhttp3.MediaType.parse("application/json")))
            .build();

        try (Response response = httpClient.newCall(request).execute()) {
            int status = response.code();
            ResponseBody body = response.body();
            if (!response.isSuccessful() || status == 207) {
                // In the case of a multi-status response (207), it likely means that one or more
                // events were rejected. In such a case, the error is actually contained in
                // the normal response body.
                throw new IOException(body.string());
            }

            log.log(INFO, "Sent " + events.size() + " events successfully.");
        }
    }
}

File: analytics/metrics-platform/src/main/java/org/wikimedia/metricsplatform/EventSenderDefault.kt
package org.wikimedia.metricsplatform

import okhttp3.OkHttpClient
import okhttp3.RequestBody
import okhttp3.ResponseBody
import org.wikimedia.metricsplatform.event.EventProcessed

class EventSenderDefault(private val gson: com.google.gson.Gson,
                         private val httpClient: OkHttpClient
) : EventSender {
    override fun sendEvents(baseUri: java.net.URL, events: List<EventProcessed>) {
        val request = okhttp3.Request.Builder()
            .url(baseUri)
            .header("Accept", "application/json")
            .header(
                "User-Agent",
                "Metrics Platform Client/Java " + MetricsClient.METRICS_PLATFORM_LIBRARY_VERSION
            )
            .post(RequestBody.create(gson.toJson(events), parse.parse("application/json")))
            .build()

        httpClient.newCall(request).execute().use { response ->
            val status: kotlin.Int = response.code
            val body: ResponseBody? = response.body
            if (!response.isSuccessful || status == 207) {
                // In the case of a multi-status response (207), it likely means that one or more
                // events were rejected. In such a case, the error is actually contained in
                // the normal response body.
                throw java.io.IOException(body.string())
            }
            //log.log(java.util.logging.Level.INFO, "Sent " + events.size + " events successfully.")
        }
    }
}
